File: jishu.py
import RPi.GPIO as GPIO
from time import sleep
from threading import Thread

GPIO.setmode(GPIO.BCM)
DIO = 4
SCK = 3
RCK = 2
table = [0xc0, 0xf9, 0xa4, 0xb0, 0x99, 0x92, 0x82, 0xf8, 0x80, 0x90, 0x88, 0x83, 0xc6, 0xa1, 0x86, 0x8e]


def write(pinNumber, state):
    GPIO.setup(pinNumber, GPIO.OUT)
    GPIO.output(pinNumber, bool(state))


def destroy():
    write(18, 0)
    print "stop"


def blink(pinNumber):
    write(pinNumber, 1)
    write(pinNumber, 0)


def shiftOut(indata):
    for i in range(8):
        write(DIO, indata & (1 << (7 - i)))
        blink(SCK)


def xianshi(numbers):
    numbers = numbers[::-1]
    for i in range(len(numbers)):
        num = int(numbers[i])
        shiftOut(1 << (7 - i))
        shiftOut(table[num])
        blink(RCK)
        sleep(0.002)

mynumber = '0'


def loop():
    i = 1      
    global mynumber
    while True:
        mynumber = str(i)
        i += 1
        sleep(0.1)
        if i % 10 == 0:
            write(18, 1)
            sleep(0.0001)
            write(18, 0)


def displayThread():
    while True:
        xianshi(mynumber)


if __name__ == '__main__':
    t = Thread(target=displayThread)
    t.setDaemon(True)
    t.start()
    try:
        loop()
    except KeyboardInterrupt:
        global mynumber
        mynumber = '0'
        sleep(0.02)
        destroy()

File: jishu.c
#include <stdio.h>
#include <wiringPi.h>
int DIO = 7, SCK = 9, RCK = 8;
unsigned char t[] = {0xc0, 0xf9, 0xa4, 0xb0, 0x99, 0x92, 0x82, 0xf8, 0x80, 0x90};
void setup() {
    wiringPiSetup();
    pinMode(DIO, OUTPUT); pinMode(SCK, OUTPUT); pinMode(RCK, OUTPUT);
    pinMode(1, OUTPUT);
}
void blink(int pin){
  digitalWrite(pin, 1);
  digitalWrite(pin, 0);
}
void shiftOut(long indata){
    int i;
    for(i=0;i<8;i++){
        digitalWrite(DIO, indata & (1 << (7 - i)));
        blink(SCK);
        nanosleep(0,10000);
    }
}
long num=1;
void loop(){
    int i=0;
    long b=num;
    while(b>0){
        shiftOut(1 << (7 - i));
        shiftOut(t[b%10]);
        blink(RCK);
        b/=10;
        i++;
    }
}

int main()
{
    setup();
    int i=0;
    while(1){
        i++;
        if(i==100){
            //printf("%ld\n",num);
            num++;
            i=0;
            if(num%100==0){
                digitalWrite(1, 1);
                delay(1);
                digitalWrite(1, 0);
            }
        }
        loop();
    }
}
